doc/design/service-discovery.md

```diff
@@ -0,0 +1,228 @@
+
+* Static discovery shall be supported
+  * Having RouDi read in a config file, which has
+    * All the topics listed
+    * Wiring pub/sub information
+  * During init phase pub/sub are connected
+* Create abstract interface for `ServiceRegistry`
+  * Add new class to `StaticServiceRegistry::StaticServiceRegistry(std::map<CaproIdString_t, instance_t> map)`
+    * No notification-based callback possibilty
+    * Just static, polling with `FindService()` possible
+
+#### Sychronous, one-shot RPC (Polling)
+
+##### Alternative A: Request/ response feature usage
+
+* Re-use [request/ response feature](https://github.com/eclipse-iceoryx/iceoryx/issues/27)
+
+Pro:
+
+* Re-use building blocks to avoid code duplication
+* Dogfooding
+* Fast due to shared memory
+
+Con:
+
+* Shared memory approach might have overhead compared for a simple RPC call
+
+##### Alternative B: IPC channel usage
+
+Stick with IPC channel
+
+Pro:
+
+* No code changes needed
+
+Con:
+
+* No userspace solution
+  * IPC channel uses OS functions
+* Upper frame limit may lead to segregation into several frames
+
+#### Event-based notification
+
+##### Alternative A: Shared memory + IPC channel
+
+* Send ID's + ABA-counter/ServiceDiscoveryChangeCounter over IPC channel
+* Data transfer over shared memory
+  * Basically putting Service Registry to shared memory so everyone can access it
+
+Pro:
+
+* Fast data transfer
+
+Contra:
+
+* More complex than todays solution
+* In an optimal case IPC channel communication shall only be used for startup
+
+##### Alternative B: Built-in topic based on `InterfacePort`'s
+
+* Built-in topic approach based on `InterfacePort`'s
+  * Sniff and intercept `CaproMessage`
+* Also see
+  * [overview diagram](diagrams/service-discovery-sequence-diagram.puml)
+  * [sequence diagram](diagrams/service-discovery-overview.puml)
+
+Pro:
+
+* Close to DDS mechanism
+* `InterfacePort` already available
+* Synergies with gateway mechanisms
+* Gateways can benefit from event notification mechanism
+  * No need for a polling-loop anymore
+
+Con:
+
+* Overhead, `Listner` will wake up on every `CaproMessage`
+* What will RouDi do if he runs out of memory?
+  * Dimensioning according to max values is not optimal (MAX_INTERFACE_CAPRO_FIFO_SIZE)
+  * Presumably lots of memory needed, e.g. during startup phase when lots of apps will do discovery
+  * A safety-certified middleware based on iceoryx would not use the dynamic discovery feature
+
+Remark:
+
+* Need to filter out your own `CaproMessageType::FIND` requests
+* Extend `dispatchCaproMessage()` with condition variable and notification mechanism
+
+##### Alternative C: Custom thread
+
+* Create a custom thread in `PoshRuntime`, which polls for specific discovery information using request/response
+
+Pro:
+
+* More flexible
+* Re-use building blocks
+
+Con:
+
+* Not needed by all users per default
+* Polling leads to overhead
+
+### Decision
+
+* Sychronous, one-shot RPC (Polling): Alternative A
```

Comment:
How did you come to this decision because I do not agree?
I would use for both push/pull alternative A which would spare us a lot of code.

The user just must be able to attach the service registry to a listener/waitset and we are fine, which can maybe abstracted away in the posh runtime.

Reply:
No final decision was done yet, I'll add this section once we've finished the discussion.

Reply:
Added final decision